File: dotnet/src/Envdoctor/Detectors/WeakSecretDetector.cs
using Envdoctor.Models;
using static Envdoctor.Detectors.DetectorHelpers;

namespace Envdoctor.Detectors;

/// Weak/placeholder secret detector. Only inspects definitions in actual
/// environment files, never `.env.example`.
public sealed class WeakSecretDetector : IDetector
{
    private static readonly HashSet<string> Blocklist = new(StringComparer.Ordinal)
    {
        "",
        "changeme",
        "password",
        "password123",
        "secret",
        "secret123",
        "token",
        "key",
        "apikey",
        "api_key",
        "test",
        "testing",
        "12345678",
        "123456789",
        "your_secret",
        "your_token",
        "your_api_key",
        "your_password",
        "example",
        "dummy",
        "foo",
        "bar",
        "admin",
        "default",
        "null",
        "undefined",
    };

    public string Id => "weak-secret";
    public string Name => "weak-secret";
    public string Description => "A secret-looking variable in an environment file has a weak or placeholder value.";

    private static bool IsWeakSecret(string value)
    {
        var trimmed = value.Trim();
        if (trimmed.Length == 0)
            return false;
        if (Blocklist.Contains(trimmed.ToLowerInvariant()))
            return true;
        return trimmed.Length < 8;
    }

    public List<Finding> Detect(IndexedModel index)
    {
        var findings = new List<Finding>();

        var entries = index.EnvDefinitions
            .OrderBy(kv => DefSortKey(kv.Value), Comparer<(string Path, int Line)>.Create(CompareKeys))
            .ThenBy(kv => kv.Key, StringComparer.Ordinal);

        foreach (var (name, defs) in entries)
        {
            foreach (var def in defs)
            {
                if (!def.IsSecret)
                    continue;
                if (def.Value is null || !IsWeakSecret(def.Value))
                    continue;
                var location = def.Origin.Line is { } line
                    ? $"{def.Origin.FilePath}:{line}"
                    : def.Origin.FilePath;
                findings.Add(MakeFinding(
                    "weak-secret",
                    Severity.Warning,
                    name,
                    $"{name} has a weak or placeholder value in {location}",
                    new List<Origin> { def.Origin.Clone() }));
            }
        }

        return findings;
    }
}

File: swift/Sources/EnvdoctorKit/Detectors/WeakSecretDetector.swift
import Foundation

/// Weak/placeholder secret detector. Only inspects definitions in actual
/// environment files, never `.env.example`.
public struct WeakSecretDetector: Detector {
    public let id = "weak-secret"
    public let name = "weak-secret"
    public let description = "A secret-looking variable in an environment file has a weak or placeholder value."

    private static let blocklist: Set<String> = [
        "", "changeme", "password", "password123", "secret", "secret123", "token",
        "key", "apikey", "api_key", "test", "testing", "12345678", "123456789",
        "1234567890", "your_secret", "your_token", "your_api_key", "your_password",
        "example", "dummy", "foo", "bar", "admin", "default", "null", "undefined",
    ]

    public init() {}

    private func isWeakSecret(_ value: String?) -> Bool {
        guard let value else { return false }
        let trimmed = value.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty { return false }
        if WeakSecretDetector.blocklist.contains(trimmed.lowercased()) { return true }
        if trimmed.count < 8 { return true }
        return false
    }

    public func detect(_ index: IndexedModel) -> [Finding] {
        var findings: [Finding] = []
        for (name, defs) in index.envDefinitions {
            for def in defs {
                if !def.isSecret { continue }
                if !isWeakSecret(def.value) { continue }
                let location = def.origin.line.map { "\(def.origin.filePath):\($0)" } ?? def.origin.filePath
                findings.append(makeFinding(
                    "weak-secret", .warning, name,
                    "\(name) has a weak or placeholder value in \(location)",
                    [def.origin]
                ))
            }
        }
        return findings
    }
}
